pilot/pkg/model/push_context.go

```diff
@@ -635,9 +676,96 @@ func (ps *PushContext) InitContext(env *Environment) error {
 	if err = ps.initSidecarScopes(env); err != nil {
 		return err
 	}
+	return err
+}
+
+func (ps *PushContext) updateFromExistingContext(env *Environment, oldPushContext *PushContext, updateReq *UpdateRequest) error {
+
+	var err error
+	servicesChanged := false
+	virtualServicesChanged := false
+	destinationRulesChanged := false
+	authZChanged := false
+	envoyFiltersChanged := false
+	sidecarsChanged := false
+
+	for k := range updateReq.ConfigTypesUpdated {
+		switch k {
+		case ServiceEntry.Type:
+			servicesChanged = true
+		case DestinationRule.Type:
+			destinationRulesChanged = true
+		case VirtualService.Type:
+			virtualServicesChanged = true
+		case Sidecar.Type:
+			sidecarsChanged = true
+		case EnvoyFilter.Type:
+			envoyFiltersChanged = true
+		case ServiceRoleBinding.Type, ServiceRole.Type, AuthorizationPolicy.Type:
+			authZChanged = true
+		}
+	}
 
-	ps.initDone = true
-	return nil
+	if servicesChanged {
+		// Services have changed. initialize service registry
+		if err = ps.initServiceRegistry(env); err != nil {
+			return err
+		}
+	} else {
+		// copy all the relevant data structures from old push context
+		ps.privateServicesByNamespace = oldPushContext.privateServicesByNamespace
+		ps.publicServices = oldPushContext.publicServices
+		ps.ServiceAccounts = oldPushContext.ServiceAccounts
+		ps.ServiceByHostnameAndNamespace = oldPushContext.ServiceByHostnameAndNamespace
+	}
+
+	if virtualServicesChanged {
+		if err = ps.initVirtualServices(env); err != nil {
+			return err
+		}
+	} else {
+		ps.publicVirtualServices = oldPushContext.publicVirtualServices
+		ps.privateVirtualServicesByNamespace = oldPushContext.privateVirtualServicesByNamespace
+	}
+
+	if destinationRulesChanged {
+		if err = ps.initDestinationRules(env); err != nil {
+			return err
+		}
+	} else {
+		ps.namespaceLocalDestRules = oldPushContext.namespaceLocalDestRules
+		ps.allExportedDestRules = oldPushContext.allExportedDestRules
+		ps.namespaceExportedDestRules = oldPushContext.namespaceExportedDestRules
+	}
+
+	if authZChanged {
+		if err = ps.initAuthorizationPolicies(env); err != nil {
+			rbacLog.Errorf("failed to initialize authorization policies: %v", err)
+			return err
+		}
+	} else {
+		ps.AuthzPolicies = oldPushContext.AuthzPolicies
+	}
+
+	if envoyFiltersChanged {
+		if err = ps.initEnvoyFilters(env); err != nil {
+			return err
+		}
+	} else {
+		ps.envoyFiltersByNamespace = oldPushContext.envoyFiltersByNamespace
+	}
+
+	// Must be initialized in the end
+	// Sidecars need to be updated if services, virtual services, destination rules, or the sidecar configs change
+	if servicesChanged || virtualServicesChanged || destinationRulesChanged || sidecarsChanged {
```

Comment:
Because sidecar scope imports virtual services and destination rules.. In cluster.go, you do push.DestinationRules(proxy).. which in turn calls proxy.SidecarScope.DestinationRules()

Reply:
btw, @ramaraochavali this is the reason I created those booleans. The if condition for sidecar scope regeneration will be a very long one checking for multiple keys.

Reply:
>Because sidecar scope imports virtual services and destination rules.

Noticed that, make sense